src/Hamiltonians/HOCartesianEnergyConserved.jl

```diff
@@ -0,0 +1,378 @@
+"""
+    four_oscillator_integral_general(i, j, k, l; max_level = typemax(Int))
+
+Integral of four one-dimensional harmonic oscillator functions, 
+```math
+    \\mathcal{I}(i,j,k,l) = \\int_{-\\infty}^\\infty dx \\, 
+    \\phi_i(x) \\phi_j(x) \\phi_k(x) \\phi_l(x)
+```
+Indices `i,j,k,l` start at `0` for the groundstate.
+
+This integral has a closed form in terms of the hypergeometric ``_{3}F_2`` function, 
+and is non-zero unless ``i+j+k+l`` is odd. See e.g. 
+[Titchmarsh 1948](https://doi.org/10.1112/jlms/s1-23.1.15).
+Used in [`HOCartesianEnergyConserved`](@ref).
+"""
+function four_oscillator_integral_general(i, j, k, l; max_level = typemax(Int))
+    all(0 .≤ (i, j, k, l) .≤ max_level) || return 0.0
+    iseven(i + j + k + l) || return 0.0
+
+    # enforce correct symmetry and k ≥ l
+    i, j, k, l = sort(SVector(i, j, k, l); rev = true)
+    
+    a = i + j - k + l + 1    
+    b = i - j + k - l + 1
+    c = -i + j + k - l + 1
+    d = -i - j + k - l + 1
+    
+    p = sqrt(2 * gamma(i + 1) * gamma(j + 1) * gamma(k + 1) * gamma(l + 1)) * pi^2
+    q = gamma(a/2) * gamma(b/2) * gamma(c/2) * gamma(k + 1) / gamma(k - l + 1)
+    
+    f = _₃F₂(float(-l), b/2, c/2, float(1 + k - l), d/2, 1.0)
+    
+    return f * q / p
+end
+
+"""
+    largest_two_point_interval(i, j, [v=1,] w) -> range
+
+For two points `i` and `j` on a range `v:w`, find the largest subinterval of sites 
+such that moving `i` to one of those sites and moving `j` by an equal but opposite
+amount leaves both within `v:w`.
+
+`i` and `j` may be `Rational`s, but `v` and `w` must be `Integer`s.
+"""
+function largest_two_point_interval(i::Int, j::Int, v::Int, w::Int)
+    i in v:w && j in v:w || throw(ArgumentError("interval out of bounds: i=$i, j=$j, v=$v, w=$w"))
+    left_edge, right_edge = two_point_interval_bounds(i, j, v, w)
+    return left_edge:right_edge
+end
+function largest_two_point_interval(i, j, v::Int, w::Int)
+    v ≤ i ≤ w && v ≤ j ≤ w || throw(ArgumentError("interval out of bounds: i=$i, j=$j, v=$v, w=$w"))
+    left_edge, right_edge = two_point_interval_bounds(i, j, v, w)
+    return ceil(Int, left_edge):floor(Int, right_edge)
+end
```

Comment:
Is this method needed? I noticed that the tests don't touch it:
[coveralls](https://coveralls.io/builds/59714810/source?filename=src%2FHamiltonians%2FHOCartesianEnergyConserved.jl#L50)